src/channel-qvm.lisp

```diff
@@ -0,0 +1,374 @@
+(defclass noise-rule ()
+  ((noise-predicate
+    :initarg :noise-predicate
+    :accessor noise-predicate
+    :initform nil 
+    :documentation "a function that is true when an instruciton is matched, false otherwise")
+   (operation-elements
+    :initarg :operation-elements
+    :accessor operation-elements
+    :initform ()  
+    :documentation "operation elements (kraus operators) to apply when the predicate is satisfied")))
+
+
+(defun make-noise-rule (predicate operation-elements)
+  "Returns a noise rule with the specified prediate and operation elements. "
+  (make-instance 'noise-rule :noise-predicate predicate :operation-elements operation-elements))
+
+
+;; A noise model consists of a set of noise rules and readout-povms. The noise rules define quantum channels and where they should be
+;; applied,  and the readout-povms describe the noisy readout probabilities. 
+(defclass noise-model ()
+  ((noise-rules
+    :initarg :noise-rules
+    :accessor noise-rules
+    :initform nil
+    :documentation "A list of noise-rules (predicate channel pairs)")
+   (readout-povms
+    :initarg :readout-povms
+    :accessor readout-povms
+    :initform (make-hash-table :test 'eql)
+    :documentation "A map of qubit to readout noise assignment probabilities. These are treated differently from noise channels (for now).")))
+
+
+(defun make-noise-model (noise-rules)
+  "Returns a noise model with noise-rules."
+  (make-instance 'noise-model :noise-rules noise-rules))
+
+
+(defmethod add-noise-models (nm1 nm2)
+  " Combines two noise models by concatenating their rules. This way, either rule can be matched by an instruction. "
+  (make-noise-model (append (noise-rules nm1) (noise-rules nm2))))
+
+
+(defmethod multiply-noise-models (nm1 nm2)
+  "Combines two noise models in a way such that both set of rules, or one and not the other could be matched by an instruction."
+  (let ((nrs (loop :for rule1 in (noise-rules nm1) 
+                   :append (loop :for rule2 in (noise-rules nm2)
+                                 :collect (make-noise-rule 
+                                           (predicate-and (noise-predicate rule1) (noise-predicate rule2)) 
+                                           (list (operation-elements rule1) (operation-elements rule2)))
+                                 :collect (make-noise-rule 
+                                           (predicate-and (noise-predicate rule1) (predicate-not (noise-predicate rule2))) 
+                                           (operation-elements rule1))
+                                 :collect (make-noise-rule 
+                                           (predicate-and (predicate-not (noise-predicate rule1)) (noise-predicate rule1)) 
+                                           (operation-elements rule2))))))
+    (make-noise-model nrs)))
+
+
+;; The following are examples of predicates! 
+(defun match-strict-qubits (qubits)
```

Comment:
Depending on how often and in which manner this will get used, you might have a preference for `(&rest qubits)` here. This would allow you to write `(match-strict-qubits 0 1)` rather than `(match-strict-qubits (list 0 1))`. If you did want to apply _that_ to a list of qubits, then you could `(apply #'match-strict-qubits (list 0 1))`